storage/backup/backup-cli/src/lib.rs

```diff
@@ -83,3 +88,145 @@ async fn get_proof_and_write(
         .await?;
     Ok(file)
 }
+
+pub fn restore_account_state<P, I>(version: u64, root_hash: HashValue, db_dir: P, iter: I)
+where
+    P: AsRef<Path> + Clone,
+    I: Iterator<Item = Result<(FileHandle, FileHandle)>>,
+{
+    let libradb =
+        LibraDB::open(db_dir, false /* read_only */, None /* pruner */).expect("DB should open.");
+    let chunk_and_proofs = iter.map(|res| {
+        let (account_state_file, proof_file) = res.expect("Input iter yielded error.");
+        let accounts = read_account_state_chunk(account_state_file)
+            .expect("Failed to read account state file.");
+        let proof = read_proof(proof_file).expect("Failed to read proof file.");
+
+        (accounts, proof)
+    });
+
+    libradb
+        .restore_account_state(chunk_and_proofs, version, root_hash)
+        .expect("Failed to restore account state.");
+}
+
+fn read_account_state_chunk(file: FileHandle) -> Result<Vec<(HashValue, AccountStateBlob)>> {
+    let content = read_file(file)?;
+
+    let mut chunk = vec![];
+    let mut reader = std::io::Cursor::new(content);
+    loop {
+        let mut buf = [0u8; HashValue::LENGTH];
+        if reader.read_exact(&mut buf).is_err() {
+            break;
+        }
+        let key = HashValue::new(buf);
+
+        let len = reader.read_u32::<LittleEndian>()?;
+        let mut buf = vec![0u8; len as usize];
+        reader.read_exact(&mut buf)?;
+        let blob = AccountStateBlob::from(buf);
+
+        chunk.push((key, blob));
+    }
+
+    Ok(chunk)
+}
+
+fn read_proof(file: FileHandle) -> Result<SparseMerkleRangeProof> {
+    let content = read_file(file)?;
+    let proof = lcs::from_bytes(&content)?;
+    Ok(proof)
+}
+
+fn read_file(file: FileHandle) -> Result<Vec<u8>> {
+    let mut content = vec![];
+    for bytes_res in block_on_stream(LocalStorage::read_file_content(&file)) {
+        content.extend(bytes_res?);
+    }
+    Ok(content)
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        adapter::local_storage::LocalStorage, backup_account_state, restore_account_state,
+    };
+    use libra_config::config::NodeConfig;
+    use libra_proptest_helpers::ValueGenerator;
+    use libra_temppath::TempPath;
+    use libra_types::transaction::PRE_GENESIS_VERSION;
+    use libradb::{test_helper::arb_blocks_to_commit, LibraDB};
+    use std::sync::Arc;
+    use storage_client::StorageReadServiceClient;
+    use storage_interface::{DbReader, DbWriter};
+    use storage_service::start_storage_service_with_db;
+
+    fn tmp_db_empty() -> (TempPath, LibraDB) {
+        let tmpdir = TempPath::new();
+        let db = LibraDB::new_for_test(&tmpdir);
+
+        (tmpdir, db)
+    }
+
+    fn tmp_db_with_random_content() -> (TempPath, LibraDB) {
+        let (tmpdir, db) = tmp_db_empty();
+        let mut cur_ver = 0;
+        for (txns_to_commit, ledger_info_with_sigs) in
+            ValueGenerator::new().generate(arb_blocks_to_commit())
```

Comment:
i guess we could also just write a proptest?

Reply:
looks cleaner like this, and it's not the point to test against different db content, just the compatibility of the backup and restore cli's